src/cockpit/polkit.py

```diff
@@ -0,0 +1,154 @@
+        try:
+            await self._communicate(process)
+        except asyncio.CancelledError:
+            logger.debug('Cancelled authentication')
+            process.terminate()
+        finally:
+            res = await process.wait()
+            logger.debug('helper exited with code %i', res)
+
+    async def _communicate(self, process: asyncio.subprocess.Process) -> None:
+        assert process.stdin
+        assert process.stdout
+
+        messages: List[str] = []
+
+        async for line in process.stdout:
+            logger.debug('Read line from helper: %s', line)
+            command, _, value = line.strip().decode().partition(' ')
+
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if command.startswith('PAM_PROMPT'):
+                # Don't pass this to the UI if it's "Password" (the usual case),
+                # so that superuser.py uses the translated default
+                if value.startswith('Password'):
+                    value = ''
+
+                # flush out accumulated info/error messages
+                passwd = await self.responder.do_askpass('\n'.join(messages), value, '')
+                messages.clear()
+                if passwd is None:
+                    logger.debug('got PAM_PROMPT %s, but do_askpass returned None', value)
+                    raise asyncio.CancelledError('no password given')
+                logger.debug('got PAM_PROMPT %s, do_askpass returned a password', value)
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif command in ('PAM_TEXT_INFO', 'PAM_ERROR'):
+                messages.append(value)
+            elif command == 'SUCCESS':
+                logger.info('Authentication succeeded')
+                break
+            elif command == 'FAILURE':
+                logger.warning('Authentication failed')
+                break
+            else:
+                logger.warning('Unknown line from helper, aborting: %s', line)
+                process.terminate()
+                break
+
+
+class PolkitAgent:
+    """Register polkit agent when required
+
+    Use this as a context manager to ensure that the agent gets unregistered again.
+    """
+    def __init__(self, responder: InteractionResponder):
+        self.responder = responder
+        self.agent_slot = None
+
+    async def __aenter__(self):
+        try:
+            self.system_bus = bus.Bus.default_system()
+        except OSError as e:
+            logger.warning('cannot connect to system bus, not registering polkit agent: %s', e)
+            return self
+
+        try:
+            # may refine that with a D-Bus call to logind
+            self.subject = ('unix-session', {'session-id': Variant(os.environ['XDG_SESSION_ID'], 's')})
```

Comment:
The `'s'` is optional, but it doesn't hurt, either.

Reply:
I prefer to keep it, to clarify the D-Bus type.